xdem/coreg.py

```diff
@@ -1063,3 +1081,604 @@ def coregister(reference_raster: Union[str, gu.georaster.Raster], to_be_aligned_
+        # Make a new DEM which will be modified inplace
+        aligned_dem = tba_dem.copy()
+
+        # Calculate slope and aspect maps from the reference DEM
+        slope, aspect = calculate_slope_and_aspect(ref_dem)
+
+        # Make index grids for the east and north dimensions
+        east_grid = np.arange(ref_dem.shape[1])
+        north_grid = np.arange(ref_dem.shape[0])
+
+        # Make a function to estimate the aligned DEM (used to construct an offset DEM)
+        elevation_function = scipy.interpolate.RectBivariateSpline(x=north_grid, y=east_grid,
+                                                                   z=np.where(np.isnan(aligned_dem), -9999, aligned_dem))
+        # Make a function to estimate nodata gaps in the aligned DEM (used to fix the estimated offset DEM)
+        nodata_function = scipy.interpolate.RectBivariateSpline(x=north_grid, y=east_grid, z=np.isnan(aligned_dem))
+        # Initialise east and north pixel offset variables (these will be incremented up and down)
+        offset_east, offset_north, bias = 0.0, 0.0, 0.0
+
+        # Iteratively run the analysis until the maximum iterations or until the error gets low enough
+        for i in trange(self.max_iterations, disable=not verbose, desc="Iteratively correcting dataset"):
+
+            # Calculate the elevation difference and the residual (NMAD) between them.
+            elevation_difference = ref_dem - aligned_dem
+            bias = np.nanmedian(elevation_difference)
+            # Correct potential biases
+            elevation_difference -= bias
+
+            nmad = xdem.spatial_tools.nmad(elevation_difference)
+
+            assert ~np.isnan(nmad), (offset_east, offset_north)
+
+            # Stop if the NMAD is low and a few iterations have been made
+            if i > 5 and nmad < self.error_threshold:
+                if verbose:
+                    print(f"NMAD went below the error threshold of {self.error_threshold}")
+                break
+
+            # Estimate the horizontal shift from the implementation by Nuth and Kääb (2011)
+            east_diff, north_diff, _ = get_horizontal_shift(  # type: ignore
+                elevation_difference=elevation_difference,
+                slope=slope,
+                aspect=aspect
+            )
+            # Increment the offsets with the overall offset
+            offset_east += east_diff
+            offset_north += north_diff
+
+            # Calculate new elevations from the offset x- and y-coordinates
+            new_elevation = elevation_function(y=east_grid + offset_east, x=north_grid - offset_north)
+
+            # Set NaNs where NaNs were in the original data
+            new_nans = nodata_function(y=east_grid + offset_east, x=north_grid - offset_north)
+            new_elevation[new_nans >= 1] = np.nan
+
+            # Assign the newly calculated elevations to the aligned_dem
+            aligned_dem = new_elevation
+
```

Comment:
Mmmh, interesting point...
Regarding point 1, I believe that once Coreg is well setup, most users won't care at all about subfunctions etc. With the suggested documentation, I find it much easier to use than having to look at individual functions, which might possibly have different structures etc. So that's rather a no on this point.
Regarding point 3, I think it's good to have flexibility and allowing to re-use the functions if possible. On the other hand, most users really won't dive that far down in the functionalities. The question is, will we need to access these functions ourselves? Right now I can't think of a test case. If so, I think we can still work around it by calling the class and running the _fit_func and _apply_func functions and reading the output in _meta?
Right now, the only issue I see is that these functions only work with raster data, no point cloud. If we decide to extract the base functions out of the class, I think they should at least accept 3-D and 2-D arrays.